crates/ruma-events/src/room/message/content_serde.rs

```diff
@@ -76,6 +77,64 @@ impl<'de> Deserialize<'de> for MessageType {
     }
 }
 
+/// Helper struct to determine the itemtype from a `serde_json::value::RawValue`
+#[derive(Debug, Deserialize)]
+struct ItemTypeDeHelper {
+    /// The item type field
+    itemtype: String,
+}
+
+impl<'de> Deserialize<'de> for GalleryItemType {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let json = Box::<RawJsonValue>::deserialize(deserializer)?;
+        let ItemTypeDeHelper { itemtype } = from_raw_json_value(&json)?;
+
+        Ok(match itemtype.as_ref() {
+            "m.audio" => Self::Audio(from_raw_json_value(&json)?),
+            "m.file" => Self::File(from_raw_json_value(&json)?),
+            "m.image" => Self::Image(from_raw_json_value(&json)?),
+            "m.video" => Self::Video(from_raw_json_value(&json)?),
+            _ => Self::_Custom(from_raw_json_value(&json)?),
+        })
+    }
+}
+
+impl Serialize for GalleryItemType {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut map = match self {
+            GalleryItemType::Audio(content) => {
+                serde_json::to_value(content).map_err(serde::ser::Error::custom)?
+            }
+            GalleryItemType::File(content) => {
+                serde_json::to_value(content).map_err(serde::ser::Error::custom)?
+            }
+            GalleryItemType::Image(content) => {
+                serde_json::to_value(content).map_err(serde::ser::Error::custom)?
+            }
+            GalleryItemType::Video(content) => {
+                serde_json::to_value(content).map_err(serde::ser::Error::custom)?
+            }
+            GalleryItemType::_Custom(content) => {
+                serde_json::to_value(content).map_err(serde::ser::Error::custom)?
+            }
+        }
+        .as_object()
+        .cloned()
+        .unwrap_or_default();
+
+        map.insert("itemtype".to_owned(), JsonValue::String(self.itemtype().to_owned()));
+        map.remove("msgtype");
+
+        map.serialize(serializer)
+    }
+}
```

Comment:
This sort of sucks. I wanted to avoid having to duplicate all the content types but still be able to serialize msgtype as itemtype. Curious if anyone has better ideas on how to achieve this.

Reply:
Is this what you wanted?
```suggestion
impl Serialize for GalleryItemType {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut map = match self {
            GalleryItemType::Audio(content) |
            GalleryItemType::File(content) |
            GalleryItemType::Image(content) |
            GalleryItemType::Video(content) |
            GalleryItemType::_Custom(content) => {
                serde_json::to_value(content).map_err(serde::ser::Error::custom)?
            }
        }
        .as_object()
        .cloned()
        .unwrap_or_default();

        map.insert("itemtype".to_owned(), JsonValue::String(self.itemtype().to_owned()));
        map.remove("msgtype");

        map.serialize(serializer)
    }
}
```

Reply:
No, actually not (and I can't seem to get this to compile). What I meant is it feels odd to serialize the content into a map and then serialize that again after modification. Maybe it's not as problematic.

Reply:
Maybe instead of having a `#[serde(tag = "msgtype", rename = "<msgtype>")]` on `*MessageEventContent`, we could have it on the enums, i.e.:

```rust
#[derive(Clone, Debug, Serialize)]
#[serde(tag = "msgtype")]
pub enum MessageType {
    /// An audio message.
    #[serde(rename = "m.audio")]
    Audio(AudioMessageEventContent),

    // …
}
```

That requires to duplicate the `msgtype`/`itemtype` in both enums, but it shouldn't require a custom `Serialize` implementation. We also have to check that it doesn't break expectations elsewhere.

Reply:
Oh wow, this is a great idea and seems to work fine from what I can tell.